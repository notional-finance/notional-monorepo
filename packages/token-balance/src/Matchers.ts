import { TokenBalance } from './tokens/TokenBalance';

export const tokenBalanceMatchers = {
  toEqTB(r: TokenBalance, a: TokenBalance) {
    try {
      r.isMatch(a);
      return {
        pass: r.eq(a),
        message: `expected ${r.toExactString()} == ${a.toExactString()}`,
      };
    } catch (e) {
      return {
        pass: false,
        message: (e as TypeError).toString(),
      };
    }
  },
  toLtTB(r: TokenBalance, a: TokenBalance) {
    try {
      r.isMatch(a);
      return {
        pass: r.lt(a),
        message: `expected ${r.toExactString()} < ${a.toExactString()}`,
      };
    } catch (e) {
      return {
        pass: false,
        message: (e as TypeError).toString(),
      };
    }
  },
  toLtEqTB(r: TokenBalance, a: TokenBalance) {
    try {
      r.isMatch(a);
      return {
        pass: r.lte(a),
        message: `expected ${r.toExactString()} <= ${a.toExactString()}`,
      };
    } catch (e) {
      return {
        pass: false,
        message: (e as TypeError).toString(),
      };
    }
  },
  toGtTB(r: TokenBalance, a: TokenBalance) {
    try {
      r.isMatch(a);
      return {
        pass: r.gt(a),
        message: `expected ${r.toExactString()} > ${a.toExactString()}`,
      };
    } catch (e) {
      return {
        pass: false,
        message: (e as TypeError).toString(),
      };
    }
  },
  toGtEqTB(r: TokenBalance, a: TokenBalance) {
    try {
      r.isMatch(a);
      return {
        pass: r.lte(a),
        message: `expected ${r.toExactString()} >= ${a.toExactString()}`,
      };
    } catch (e) {
      return {
        pass: false,
        message: (e as TypeError).toString(),
      };
    }
  },
  toBeApprox(r: TokenBalance, a: TokenBalance, rel = 1e-5) {
    try {
      r.isMatch(a);
      const pass = rel ? Math.abs(r.toFloat() / a.toFloat() - 1) <= rel : true;
      return {
        pass,
        message: `expected ${r.toExactString()} +/- ${rel} == ${a.toExactString()}`,
      };
    } catch (e) {
      return {
        pass: false,
        message: (e as TypeError).toString(),
      };
    }
  },
};
