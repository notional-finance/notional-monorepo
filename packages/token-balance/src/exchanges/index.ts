import MetaStable2Token from './BalancerV2/MetaStableTwoToken';

export default { MetaStable2Token };
