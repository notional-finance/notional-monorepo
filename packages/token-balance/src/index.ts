export * from './TokenRegistry';
export * from './TokenBalance';
export * from './Matchers';
