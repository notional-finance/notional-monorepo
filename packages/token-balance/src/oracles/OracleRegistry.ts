import { BigNumber, Contract, ethers } from 'ethers';
import {
  BehaviorSubject,
  combineLatest,
  map,
  of,
  Subject,
  withLatestFrom,
} from 'rxjs';
import {
  AssetRateAggregatorABI,
  IAggregator,
  IAggregatorABI,
} from '@notional-finance/contracts';
import { aggregate, AggregateCall } from '@notional-finance/multicall';
import defaultOracles from './DefaultOracles';
import {
  ExchangeRate,
  Network,
  OracleDefinition,
  OracleInterface,
} from '../Definitions';
import {
  RATE_PRECISION,
  RATE_PRECISION_SQUARED,
} from '@notional-finance/sdk/config/constants';
import { getNowSeconds } from '@notional-finance/util';

type OracleSubject = BehaviorSubject<ExchangeRate | undefined>;
type AdjList = Map<string, Set<string>>;

interface OracleGraph {
  adjList: AdjList;
  oracles: Map<string, OracleDefinition[]>;
  subjects: Map<string, OracleSubject>;
}

interface OraclePath {
  key: string;
  oracleIndex: number;
  mustInvert: boolean;
}

export class OracleRegistry {
  protected static lastUpdateBlock = new Map<
    Network,
    BehaviorSubject<number>
  >();

  protected static lastUpdateTimestamp = new Map<
    Network,
    BehaviorSubject<number>
  >();

  protected static oracleGraphs = new Map<Network, OracleGraph>(
    defaultOracles.map(([n, _oracles]) => {
      // Initializes the oracle graph using the default oracles
      const adjList = new Map<string, Set<string>>();
      const oracles = new Map<string, OracleDefinition[]>();
      const subjects = new Map<string, OracleSubject>();

      _oracles.forEach((o) => {
        this.addOracle(o, adjList, oracles, subjects);
      });

      // Reset the last update block to zero
      this.lastUpdateBlock.set(n, new BehaviorSubject<number>(0));

      return [n, { oracles, adjList, subjects }];
    })
  );

  protected static addOracle(
    o: OracleDefinition,
    adjList: AdjList,
    oracles: Map<string, OracleDefinition[]>,
    subjects: Map<string, OracleSubject>
  ) {
    const quoteToBase = adjList.get(o.quote) || new Set<string>();
    quoteToBase.add(o.base);
    adjList.set(o.quote, quoteToBase);

    const baseToQuote = adjList.get(o.base) || new Set<string>();
    baseToQuote.add(o.quote);
    adjList.set(o.base, baseToQuote);

    const key = this.getOracleKey(o);
    oracles.set(key, (oracles.get(key) || []).concat(o));

    // Index Key is always defined here
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    const index = oracles.get(key)!.length - 1;
    const indexKey = this.getOracleKey(o, index);
    subjects.set(
      indexKey,
      new BehaviorSubject<ExchangeRate | undefined>(undefined)
    );
  }

  public static getOracleKey(o: OracleDefinition, oracleIndex?: number) {
    return oracleIndex !== undefined
      ? `${o.base}/${o.quote}:${oracleIndex}`
      : `${o.base}/${o.quote}`;
  }

  public static getOracleGraph(network: Network) {
    const oracleGraph = this.oracleGraphs.get(network);
    if (!oracleGraph) throw Error(`Oracle Graph not found for ${network}`);
    return oracleGraph;
  }

  public static isManipulationResistant(oracleInterface: OracleInterface) {
    return (
      oracleInterface === OracleInterface.Chainlink ||
      oracleInterface === OracleInterface.CompoundV2_cToken ||
      oracleInterface === OracleInterface.Notional_InterestRate ||
      oracleInterface === OracleInterface.Notional_VaultOracle
    );
  }

  public static registerOracle(network: Network, oracle: OracleDefinition) {
    const {
      adjList: oracleAdjList,
      oracles,
      subjects,
    } = this.getOracleGraph(network);
    this.addOracle(oracle, oracleAdjList, oracles, subjects);
  }

  public static getAggregateCall(
    key: string,
    oracle: OracleDefinition,
    provider: ethers.providers.Provider
  ): AggregateCall {
    // All oracles are transformed into RATE_PRECISION decimal places
    const defaultScale = (
      result: BigNumber,
      validTimestamp: number = getNowSeconds()
    ): ExchangeRate => {
      const decimals = BigNumber.from(10).pow(oracle.decimalPlaces);
      return {
        rate: result.mul(RATE_PRECISION).div(decimals),
        base: oracle.base,
        quote: oracle.quote,
        validTimestamp,
      };
    };

    switch (oracle.oracleInterface) {
      case OracleInterface.Chainlink:
        return {
          key,
          target: new Contract(oracle.address, IAggregatorABI, provider),
          method: 'latestRoundData',
          args: [],
          transform: (
            r: Awaited<ReturnType<IAggregator['functions']['latestRoundData']>>
          ) => defaultScale(r.answer, r.updatedAt.toNumber()),
        };

      case OracleInterface.CompoundV2_cToken:
        return {
          key,
          target: new Contract(
            oracle.address,
            AssetRateAggregatorABI,
            provider
          ),
          method: 'getExchangeRateView',
          args: [],
          transform: (r: BigNumber) => defaultScale(r),
        };

      default:
        throw Error('Unsupported Oracle');
    }
  }

  public static getAggregateMulticallData(
    network: Network,
    provider: ethers.providers.Provider
  ): AggregateCall[] {
    const { oracles } = this.getOracleGraph(network);
    return Array.from(oracles.values()).flatMap((oracleList) =>
      oracleList.map((o, i) =>
        this.getAggregateCall(this.getOracleKey(o, i), o, provider)
      )
    );
  }

  public static async fetchOracleData(
    network: Network,
    provider: ethers.providers.Provider
  ) {
    const aggregateCall = this.getAggregateMulticallData(network, provider);
    const { subjects } = this.getOracleGraph(network);
    const { block, results } = await aggregate<Record<string, ExchangeRate>>(
      aggregateCall,
      provider,
      // This will call next() on all the update subjects
      subjects as Map<string, Subject<unknown>>
    );

    // TODO: also update timestamp here
    this.lastUpdateBlock.get(network)?.next(block.number);
    this.lastUpdateTimestamp.get(network)?.next(block.timestamp);

    return { blockNumber: block.number, results };
  }

  // public static async fetchFromCache(_cacheUrl: string) {}

  protected static breadthFirstSearch(
    base: string,
    quote: string,
    adjList: AdjList
  ) {
    if (base === quote) throw Error(`Invalid exchange rate ${base}/${quote}`);

    let path = [base];
    const queue = [path];

    while (queue.length > 0) {
      const currentPath = queue.shift();
      if (!currentPath || currentPath.length === 0) continue;

      const lastSymbol = currentPath[currentPath.length - 1];
      // If the last symbol of the path is the base then quit
      if (lastSymbol === quote) {
        path = currentPath;
        break;
      }

      // Loop into nodes linked to the last symbol
      Array.from(adjList.get(lastSymbol)?.values() || []).forEach((s) => {
        // Check if the current path includes the symbol, if it does then skip adding it
        if (!currentPath.includes(s)) queue.push([...currentPath, s]);
      });
    }

    // This ensures there are at least 2 entries in the path since base !== quote
    if (path[path.length - 1] !== quote)
      throw Error(`Path from ${base} to ${quote} not found`);

    return path;
  }

  public static findPath(
    base: string,
    quote: string,
    network: Network,
    onlyManipulationResistant = true
  ) {
    // TODO: switch on only manipulation resistant oracles
    const { adjList, oracles } = this.getOracleGraph(network);
    const path = this.breadthFirstSearch(base, quote, adjList);

    return path.reduce((p, c, i) => {
      if (i === 0) return p;

      let key = `${path[i - 1]}/${c}`;
      let mustInvert = false;
      let oracleIndex = oracles
        .get(key)
        ?.findIndex((o) =>
          onlyManipulationResistant
            ? this.isManipulationResistant(o.oracleInterface)
            : true
        );

      if (oracleIndex === undefined || oracleIndex < 0) {
        key = `${c}/${path[i - 1]}`;
        mustInvert = true;
        oracleIndex = oracles
          .get(key)
          ?.findIndex((o) =>
            onlyManipulationResistant
              ? this.isManipulationResistant(o.oracleInterface)
              : true
          );
      }

      if (oracleIndex === undefined || oracleIndex < 0)
        throw Error(`No oracle found for ${key}`);
      p.push({ key, oracleIndex, mustInvert });
      return p;
    }, [] as OraclePath[]);
  }

  /**
   * @param network
   * @param oraclePath
   * @returns an array of observables that represents the path
   */
  protected static getObservablesFromPath(
    network: Network,
    oraclePath: OraclePath[]
  ) {
    const { subjects } = this.getOracleGraph(network);
    return oraclePath.map(({ key, oracleIndex, mustInvert }) => {
      const observable = subjects.get(`${key}:${oracleIndex}`)?.asObservable();
      if (!observable)
        throw Error(`Update Subject for ${key}:${oracleIndex} not found`);

      return mustInvert
        ? observable.pipe(
            map((r: ExchangeRate | undefined) => {
              if (r) {
                return {
                  // Invert the base and quote
                  base: r.quote,
                  quote: r.base,
                  rate: RATE_PRECISION_SQUARED.div(r.rate),
                  validTimestamp: r.validTimestamp,
                };
              }

              return undefined;
            })
          )
        : observable;
    });
  }

  private static combineRates(rates: (ExchangeRate | undefined)[]) {
    if (rates.length === 0) return undefined;

    const base = rates[0]?.base;
    const quote = rates[rates.length - 1]?.quote;

    const rate = rates.reduce(
      (p, er) => (er && p ? p.mul(er.rate).div(RATE_PRECISION) : undefined),
      BigNumber.from(RATE_PRECISION) as BigNumber | undefined
    );

    const validTimestamp = Math.min(
      ...rates.map((r) => r?.validTimestamp || 0)
    );

    return rate && base && quote
      ? { base, quote, rate, validTimestamp }
      : undefined;
  }

  public static getLatestFromPath(network: Network, oraclePath: OraclePath[]) {
    const observables = this.getObservablesFromPath(network, oraclePath);
    let latestRate: ExchangeRate | undefined;

    of(1)
      .pipe(
        withLatestFrom(...observables),
        map(([, ...rates]) => this.combineRates(rates))
      )
      .forEach((v) => {
        latestRate = v;
      });

    return latestRate;
  }

  public static subscribeToPath(network: Network, oraclePath: OraclePath[]) {
    const observables = this.getObservablesFromPath(network, oraclePath);
    return combineLatest(observables).pipe(
      map((rates) => this.combineRates(rates))
    );
  }

  public static getLatestFromOracle(network: Network, oracleIndexKey: string) {
    const { subjects } = this.getOracleGraph(network);
    const s = subjects.get(oracleIndexKey);
    if (!s)
      throw Error(
        `Oracle Index Key: ${oracleIndexKey} not found on network ${network}`
      );
    return s.value;
  }

  public static subscribeToOracle(network: Network, oracleIndexKey: string) {
    const { subjects } = this.getOracleGraph(network);
    const s = subjects.get(oracleIndexKey);
    if (!s)
      throw Error(
        `Oracle Index Key: ${oracleIndexKey} not found on network ${network}`
      );
    // shareReplay ensures that the latest results trigger the observable
    return s.asObservable();
  }

  public static getLastUpdateBlock(network: Network) {
    return this.lastUpdateBlock.get(network)?.value || 0;
  }

  public static subscribeLastUpdateBlock(network: Network) {
    return this.lastUpdateBlock.get(network)?.asObservable();
  }

  public static getLastUpdateTimestamp(network: Network) {
    return this.lastUpdateTimestamp.get(network)?.value || 0;
  }

  public static subscribeLastUpdateTimestamp(network: Network) {
    return this.lastUpdateTimestamp.get(network)?.asObservable();
  }
}
