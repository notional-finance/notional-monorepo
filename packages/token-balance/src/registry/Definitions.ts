export enum Network {
  All,
  Mainnet,
  Goerli,
}

export enum TokenInterface {
  /** Standard transferrable ERC20 */
  ERC20,
  /** ERC4626 yield bearing tokens with an extension of ERC20 */
  ERC4626,
  /** Special handling for ETH */
  ETH,
  /** Special handling for WETH */
  WETH,
  /** Notional internal ERC20 balances, denominated as 8 decimals  */
  Notional_InternalERC20,
  /** Notional fCash market liquidity tokens, only held by the nToken */
  Notional_LiquidityToken,
  /** Notional fCash, associated with a maturity */
  Notional_fCash,
  /** Notional leveraged vault shares, associated with a maturity */
  Notional_VaultShare,
  /** Notional leveraged vault strategy tokens, associated with a maturity */
  Notional_StrategyToken,
  /** Non-crypto denominated fiat balance (i.e. USD, JPY, EUR) */
  FIAT,
}

export interface TokenDefinition {
  /** Address of the token */
  address: string;
  /** Network the address refers to */
  network: Network;
  /** String based token symbol */
  symbol: string;
  /** Number of native decimal places for the token */
  decimalPlaces: number;
  /** Defines the token standard this token relies on */
  tokenInterface: TokenInterface;
  /** Defines the maturity of the token, if it exists */
  maturity?: number;
}

export enum OracleInterface {
  /** Standard Chainlink IAggregator interface */
  Chainlink,
  /** Balancer V2 time weighted average price oracle */
  BalancerV2_TWAP,
  /** Compound V2 cToken exchange rate */
  CompoundV2_cToken,
  /** Notional time weighted average interest rate oracle */
  Notional_InterestRate,
  /** Notional leveraged vault price oracle */
  Notional_VaultOracle,
}

export interface OracleDefinition {
  /** Address of the oracle */
  address: string;
  /** Network the address refers to */
  network: Network;
  /** Type of oracle interface */
  oracleInterface: OracleInterface;
  /**
   * Base for the oracle, rate is quoted as 1 unit of this token.
   * Must be a valid symbol on the same network.
   */
  base: string;
  /**
   * Quote for the oracle, rate is how many tokens 1 unit of base will purchase.
   * Must be a valid symbol on the same network
   */
  quote: string;
  /** Number of decimal places the rate is quoted in */
  decimalPlaces: number;
  /** Maximum update cadence in seconds, if defined. Typically only valid for off-chain oracles */
  heartbeat?: number;
}
