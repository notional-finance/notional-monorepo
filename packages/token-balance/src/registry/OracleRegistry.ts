import defaultOracles from './DefaultOracles';

export class OracleRegistry {
  private static oracles: typeof defaultOracles;

  // generateMulticallData
  // refreshOracleData
}
