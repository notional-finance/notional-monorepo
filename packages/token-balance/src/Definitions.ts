import { BigNumber } from 'ethers';
import { BaseLiquidityPool } from './exchanges';

export enum Network {
  All = 'all',
  Mainnet = 'mainnet',
  Goerli = 'goerli',
}

export enum TokenInterface {
  /** Standard transferrable ERC20 */
  ERC20 = 'ERC20',
  /** ERC4626 yield bearing tokens with an extension of ERC20 */
  ERC4626 = 'ERC4626',
  /** Special handling for ETH */
  ETH = 'ETH',
  /** Special handling for WETH */
  WETH = 'WETH',
  /** Notional internal ERC20 balances, denominated as 8 decimals  */
  Notional_InternalERC20 = 'Notional_InternalERC20',
  /** Notional fCash market liquidity tokens, only held by the nToken */
  Notional_LiquidityToken = 'Notional_LiquidityToken',
  /** Notional fCash, associated with a maturity */
  fCash = 'fCash',
  /** Notional leveraged vault shares, associated with a maturity */
  VaultShare = 'VaultShare',
  /** Notional leveraged vault strategy tokens, associated with a maturity */
  StrategyToken = 'StrategyToken',
  /** Non-crypto denominated fiat balance (i.e. USD, JPY, EUR) */
  FIAT = 'FIAT',
}

export interface TokenDefinition {
  /** Address of the token */
  address: string;
  /** Network the address refers to */
  network: Network;
  /** String based token symbol */
  symbol: string;
  /** Number of native decimal places for the token */
  decimalPlaces: number;
  /** Defines the token standard this token relies on */
  tokenInterface: TokenInterface;
  /** Defines the maturity of the token, if it exists */
  maturity?: number;
}

export enum OracleInterface {
  /** Standard Chainlink IAggregator interface */
  Chainlink,
  /** Balancer V2 time weighted average price oracle */
  BalancerV2_TWAP,
  /** Compound V2 cToken exchange rate */
  CompoundV2_cToken,
  /** Notional time weighted average interest rate oracle */
  Notional_InterestRate,
  /** Notional leveraged vault price oracle */
  Notional_VaultOracle,
}

export interface OracleDefinition {
  /** Address of the oracle */
  address: string;
  /** Network the address refers to */
  network: Network;
  /** Type of oracle interface */
  oracleInterface: OracleInterface;
  /**
   * Base for the oracle, rate is quoted as 1 unit of this token.
   * Must be a valid symbol on the same network.
   */
  base: string;
  /**
   * Quote for the oracle, rate is how many tokens 1 unit of base will purchase.
   * Must be a valid symbol on the same network
   */
  quote: string;
  /** Number of decimal places the rate is quoted in */
  decimalPlaces: number;
  /** Maximum update cadence in seconds, if defined. Typically only valid for off-chain oracles */
  heartbeat?: number;
}

export interface ExchangeRate {
  /** Symbol of the base currency */
  base: string;
  /** Symbol of the quote currency */
  quote: string;
  /** Exchange rate in RATE_PRECISION */
  rate: BigNumber;
  /** Timestamp of the last update, can be used to check for value freshness */
  validTimestamp: number;
}

export interface PoolDefinition {
  /** Address of the pool */
  address: string;
  /** Typescript class of the pool to instantiate */
  poolClass: typeof BaseLiquidityPool;
  /** LP token definition to add to TokenRegistry */
  lpToken: TokenDefinition;
}
