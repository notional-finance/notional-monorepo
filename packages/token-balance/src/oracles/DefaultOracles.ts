import { Network, OracleDefinition, OracleInterface } from '../Definitions';

const assignDefaults = (
  list: Partial<OracleDefinition>[],
  defaultProps: Partial<OracleDefinition>
) => list.map((o) => Object.assign(o, defaultProps));

// Chainlink price feeds can be found here: https://docs.chain.link/data-feeds/price-feeds/addresses
const mainnetChainlinkUSD = assignDefaults(
  [
    {
      address: 'eth-usd.data.eth',
      heartbeat: 3600,
      base: 'ETH',
    },
    {
      address: 'btc-usd.data.eth',
      heartbeat: 3600,
      base: 'BTC',
    },
    {
      address: 'dai-usd.data.eth',
      heartbeat: 3600,
      base: 'DAI',
    },
    {
      address: 'usdc-usd.data.eth',
      heartbeat: 3600,
      base: 'USDC',
    },
    {
      address: 'usdt-usd.data.eth',
      heartbeat: 3600,
      base: 'USDT',
    },
    {
      address: 'comp-usd.data.eth',
      heartbeat: 3600,
      base: 'COMP',
    },
    {
      // NOTE: unmarked chainlink oracle stETH/USD
      address: '0xcfe54b5cd566ab89272946f602d76ea879cab4a8',
      heartbeat: 3600,
      base: 'stETH',
    },
    {
      // NOTE: custom oracle deployed by Notional
      address: '0x8770d8dEb4Bc923bf929cd260280B5F1dd69564D',
      heartbeat: 86400,
      base: 'wstETH',
    },
  ],
  { oracleInterface: OracleInterface.Chainlink, quote: 'USD', decimalPlaces: 8 }
);

const mainnetChainlinkCrypto = assignDefaults(
  [
    {
      address: 'wbtc-btc.data.eth',
      heartbeat: 86400,
      base: 'WBTC',
      quote: 'BTC',
      decimalPlaces: 8,
    },
    {
      address: 'bal-eth.data.eth',
      heartbeat: 86400,
      base: 'BAL',
      quote: 'ETH',
      decimalPlaces: 18,
    },
    {
      address: 'steth-eth.data.eth',
      heartbeat: 86400,
      base: 'stETH',
      quote: 'ETH',
      decimalPlaces: 18,
    },
  ],
  { oracleInterface: OracleInterface.Chainlink }
);

const compundV2AssetRates = assignDefaults(
  [
    {
      address: '0x8e3d447ebe244db6d28e2303bca86ef3033cfad6',
      quote: 'cETH',
      base: 'ETH',
    },
    {
      address: '0x719993e82974f5b5ea0c5eba25c260cd5af78e00',
      quote: 'cDAI',
      base: 'DAI',
    },
    {
      address: '0x612741825acedc6f88d8709319fe65bcb015c693',
      quote: 'cUSDC',
      base: 'USDC',
    },
    {
      address: '0x39d9590721331b13c8e9a42941a2b961b513e69d',
      quote: 'cWBTC',
      base: 'WBTC',
    },
  ],
  { oracleInterface: OracleInterface.CompoundV2_cToken }
);

const mainnetChainlinkFiat = assignDefaults(
  [
    'EUR',
    'AUD',
    'CAD',
    'CHF',
    'GBP',
    'JPY',
    'CNY',
    'KRW',
    'NZD',
    'BRL',
    'SGD',
    'TRY',
  ].map((base) => {
    return {
      address: `${base.toLowerCase()}-usd.data.eth`,
      base,
    };
  }),
  {
    oracleInterface: OracleInterface.Chainlink,
    heartbeat: 86400,
    quote: 'USD',
    decimalPlaces: 8,
  }
);

const assignNetwork = (
  oracles: Partial<OracleDefinition>[][],
  network: Network
): [Network, OracleDefinition[]] => [
  network,
  oracles.flat().map((o) => {
    // Adds the network to the oracle object and adds it to a mapping
    return Object.assign(o, { network }) as OracleDefinition;
  }),
];

const defaultOracles = [
  assignNetwork(
    [mainnetChainlinkCrypto, mainnetChainlinkUSD, compundV2AssetRates],
    Network.Mainnet
  ),
  assignNetwork([mainnetChainlinkFiat], Network.All),
];

export default defaultOracles;
