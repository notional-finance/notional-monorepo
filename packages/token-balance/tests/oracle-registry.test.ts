import { Network } from '../src/Definitions';
import { OracleRegistry } from '../src/oracles/OracleRegistry';

describe('Oracle Path', () => {
  it('[FORWARD] can find a path from usd => eth', () => {
    const path = OracleRegistry.findPath('USD', 'ETH', Network.Mainnet);
    expect(path.length).toBe(1);
    expect(path[0].key).toBe('ETH/USD');
    expect(path[0].mustInvert).toBe(false);
  });

  it('[REVERSE] can find a path from usd => eth', () => {
    const path = OracleRegistry.findPath('ETH', 'USD', Network.Mainnet);
    expect(path.length).toBe(1);
    expect(path[0].key).toBe('ETH/USD');
    expect(path[0].mustInvert).toBe(true);
  });

  it('[MULTIHOP] can find a path from cUSDC => cETH', () => {
    const path = OracleRegistry.findPath('cETH', 'cUSDC', Network.Mainnet);
    expect(path.length).toBe(4);
    expect(path.map((p) => p.key)).toEqual([
      'USDC/cUSDC',
      'USDC/USD',
      'ETH/USD',
      'ETH/cETH',
    ]);
    expect(path.map((p) => p.mustInvert)).toEqual([true, false, true, false]);

    const revPath = OracleRegistry.findPath('cUSDC', 'cETH', Network.Mainnet);
    expect(revPath.length).toBe(4);
    expect(revPath.map((p) => p.key)).toEqual([
      'ETH/cETH',
      'ETH/USD',
      'USDC/USD',
      'USDC/cUSDC',
    ]);
    expect(path.map((p) => p.mustInvert)).toEqual([true, false, true, false]);
  });

  it('[ERROR] throws on an unknown path', () => {
    expect(() =>
      OracleRegistry.findPath('xxxx', 'USD', Network.Mainnet)
    ).toThrowError('Path from xxxx to USD not found');
  });
});

describe('Fetch Oracle Rates', () => {
  // TODO: use a snapshot test here on a fixed block
  it('generates aggregate calls and fetches oracle rates', () => {
    // OracleRegistry.fetchOracleData(Network.Mainnet, hre.provider);
  });

  it.todo('returns the latest value from an oracle path');
  it.todo('returns undefined if the oracle path is not complete');
  it.todo('returns the latest value from a single oracle');
  it.todo('subscribes to an oracle path');
  it.todo('subscribes to a single oracle');
  it.todo('subscribes to update blocks');
});
