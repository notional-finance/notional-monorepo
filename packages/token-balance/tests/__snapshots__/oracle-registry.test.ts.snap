// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fetch Oracle Rates generates aggregate calls and fetches oracle rates 1`] = `
Object {
  "BAL/ETH:0": Object {
    "hex": "0x402268",
    "type": "BigNumber",
  },
  "BTC/USD:0": Object {
    "hex": "0x13b94404ae0a",
    "type": "BigNumber",
  },
  "COMP/USD:0": Object {
    "hex": "0x0b9e459300",
    "type": "BigNumber",
  },
  "DAI/USD:0": Object {
    "hex": "0x3b9b0a24",
    "type": "BigNumber",
  },
  "ETH/USD:0": Object {
    "hex": "0x0161be9f91d0",
    "type": "BigNumber",
  },
  "USDC/USD:0": Object {
    "hex": "0x3b9aca00",
    "type": "BigNumber",
  },
  "USDT/USD:0": Object {
    "hex": "0x3b9c50a0",
    "type": "BigNumber",
  },
  "WBTC/BTC:0": Object {
    "hex": "0x3b949464",
    "type": "BigNumber",
  },
  "cDAI/DAI:0": Object {
    "hex": "0x015220f0",
    "type": "BigNumber",
  },
  "cETH/ETH:0": Object {
    "hex": "0x013265d6",
    "type": "BigNumber",
  },
  "cUSDC/USDC:0": Object {
    "hex": "0x015af32a",
    "type": "BigNumber",
  },
  "cWBTC/WBTC:0": Object {
    "hex": "0x01324c1d",
    "type": "BigNumber",
  },
  "stETH/ETH:0": Object {
    "hex": "0x3b78ec8b",
    "type": "BigNumber",
  },
  "stETH/USD:0": Object {
    "hex": "0x0161594f6184",
    "type": "BigNumber",
  },
  "wstETH/USD:0": Object {
    "hex": "0x0187ec65e191",
    "type": "BigNumber",
  },
}
`;
